sklearn/semi_supervised/_self_training.py

```diff
@@ -0,0 +1,317 @@
+import numpy as np
+
+from ..base import BaseEstimator, clone
+from ..utils.validation import check_X_y, check_array, check_is_fitted
+from ..utils.metaestimators import if_delegate_has_method
+from ..utils import safe_mask
+from ..exceptions import ConvergenceWarning
+import warnings
+
+__all__ = ["SelfTrainingClassifier"]
+
+# Authors: Oliver Rausch   <rauscho@ethz.ch>
+#          Patrice Becker  <beckerp@ethz.ch>
+# License: BSD 3 clause
+
+
+def _validate_estimator(estimator):
+    """Make sure that an estimator implements the necessary methods."""
+    if estimator is None:
+        raise ValueError("base_classifier cannot be None!")
+
+    if not hasattr(estimator, "predict_proba"):
+        name = type(estimator).__name__
+        msg = "base_classifier ({}) should implement predict_proba!"
+        msg = msg.format(name)
+        raise ValueError(msg)
+
+
+class SelfTrainingClassifier(BaseEstimator):
+    """Self-training classifier
+
+    This class allows a given supervised classifier to function as a
+    semi-supervised classifier, allowing it to learn from unlabeled data. It
+    does this by iteratively predicting labels for the unlabeled data.
+
+    Read more in the :ref:`User Guide <self_training>`.
+
+    Parameters
+    ----------
+    base_classifier : estimator object
+        An estimator object implementing ``fit`` and ``predict_proba``.
+        Invoking the ``fit`` method will fit a clone of the passed estimator,
+        which will be stored in the ``base_classifier_`` attribute.
+
+    threshold : float, optional (default=0.75)
+        The decision threshold. If the ``base_classifier`` makes a prediction
+        with a ``predict_proba`` above this threshold, it will be added to the
+        labeled dataset.
+        Should be in [0, 1).
+
+    max_iter : int or ``None``, optional (default=10)
+        Maximum number of iterations allowed. Should be greater than or equal
+        to 0. If it is ``None``, the classifier will continue to predict labels
+        until all unlabeled samples have been labeled. In this case, be aware
+        that the fit may never terminate.
+
+    early_stopping: boolean, optional (default=True)
```

Comment:
This strategy is sometimes seen in literature ([e.g. this survey](http://ruder.io/semi-supervised/)). The motivation is not to improve prediction performance (like early stopping used in NNs) but to reduce training time (without impacting prediction performance).

For deterministic classifiers, there is no change classification performance when early stopping is enabled (if no new labels are added in iteration i, they will never change in subsequent iterations >i since we are always fitting on the same labels). The argument is similar for randomized classifiers (if no new labels have been produced for several iterations, it becomes more and more unlikely that new labels will be produced before the iteration limit is reached).

Reply:
> Why is this needed aside from `n_iter_no_change`?

I'm not sure I understand. Do you mean that having the boolean `early_stopping` is redundant?  I'd agree, but `MLPClassifier` also has a `early_stopping` boolean, so I mirrored their API.

Reply:
I think what Joel means is that `early_stopping` is redundant with `n_iter_no_change` which can be set to None to disable early stopping. That's what other estimators do (e.g. gradient boosting trees).

Reply:
Ok, I will fix it as soon as I have some time.

Reply:
I've removed the `early_stopping` parameter.